base/array.jl

```diff
@@ -677,9 +677,16 @@ function grow_to!(dest, itr, st)
 end
 
 ## Iteration ##
-function iterate(A::Array, i=1)
-    @_propagate_inbounds_meta
-    i >= length(A) + 1 ? nothing : (A[i], i+1)
+function iterate(A::Array)
+    @_inline_meta
+    isempty(A) && return nothing
+    @inbounds return (A[1], 1)
+end
+
+function iterate(A::Array, i::Int)
```

Comment:
The problem with this is that iterate is not necessarily a private interface, so if somebody passed say `iterate(A, -1)`, that could cause trouble. Perhaps a better thing to do would be to teach LLVM more about array lengths such that it can elide the boundscheck itself when the iteration protocol is used correctly.

Reply:
We could try using an unsigned type and starting the state at `i=0`. That way all inputs, other than the one we explicitly check for would be valid.

Reply:
For the record, maybe a pattern to enforce users not to break the state:

```
struct State
  i::Int
  State() = new(1)
  State(prev::State) = new(prev.i + 1)
end

State() # State(1)
State(State()) # State(2)
State(-1) # error
```

Reply:
Since we're currently asserting `i <= N && 0 < i <= N`, dropping just the duplicate test (and keeping the `i > 0 || throw(BoundsError())`) seems like it would still be a strict improvement